patterns_test.go

```diff
@@ -44,14 +44,35 @@ var TestVectors = map[string]*testVector{
 			"/ip4/1.2.3.4/tcp/1234/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
 			"/ip6/::/tcp/1234/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
 			"/ip6/::/udp/1234/utp/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
-			"/ip4/0.0.0.0/udp/1234/utp/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ"},
+			"/ip4/0.0.0.0/udp/1234/utp/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
+		},
 		Bad: []string{
 			"/ip4/1.2.3.4/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
 			"/ip6/::/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
 			"/tcp/123/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
 			"/ip6/::/udp/1234/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
 			"/ip6/::/utp/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
-			"/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ"}},
+			"/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
+		},
+	},
+	"DNS": {
+		Pattern: DNS,
+		Good:    []string{"/dnsaddr/ipfs.io", "/dns4/ipfs.io", "/dns4/libp2p.io", "/dns6/protocol.ai"},
+		Bad:     []string{"/ip4/127.0.0.1"},
+	},
+	"WebRTCDirect": {
+		Pattern: WebRTCDirect,
+		Good:    []string{"/ip4/1.2.3.4/tcp/3456/http/p2p-webrtc-direct", "/ip6/::/tcp/0/http/p2p-webrtc-direct"},
+		Bad:     []string{"/ip4/0.0.0.0", "/ip6/fc00::", "/udp/12345", "/ip6/fc00::/tcp/5523/udp/9543"},
+	},
+	"HTTP": {
+		Pattern: HTTP,
+		Good:    []string{"/ip4/1.2.3.4/http", "/dns4/libp2p.io/http", "/dns6/::/tcp/7011/http", "/ip6/fc00::/http"},
```

Comment:
I would add explicit `/dnsaddr/example.go`.

Reply:
That shouldn't match.

Reply:
That's the issue with multiaddr: there's no clear way to compose components, and the implementations dictate the norm. It turns out that js-mafmt consider this valid, see https://github.com/multiformats/js-mafmt/blob/master/README.md#api.

I assume the expectation is to default to port 80 over tcp.